src/cpp/src/speculative_decoding/speculative_decoding_eagle3_impl.cpp

```diff
@@ -0,0 +1,327 @@
+    m_tokenizer = main_model_tokenizer;
+    // for eagle model, we need to obtain hidden layer state as extra output
+    // apply transformations needed to run eagle model
+    share_embedding_weights(main_model, draft_model);
+    hidden_state_transform(main_model, hidden_layers);
+    // move the FC layer from draft model to main model
+    shift_fc_from_draft_to_main(main_model, draft_model);
+    hidden_state_transform(draft_model, { -1 });
+
+    // to create `main_pipeline` with enabled validation_mode and `draft_pipeline` with disabled validation mode
+    m_main_pipeline = std::make_shared<ContinuousBatchingForEagle3DecodingImpl>(main_model,
+                                                                               main_model_tokenizer,
+                                                                               main_model_desc.generation_config,
+                                                                               scheduler_configs.first,
+                                                                               main_device,
+                                                                               main_model_desc.properties,
+                                                                               true);
+    m_draft_pipeline = std::make_shared<ContinuousBatchingForEagle3DecodingImpl>(draft_model,
+                                                                                draft_model_tokenizer,
+                                                                                draft_model_desc.generation_config,
+                                                                                scheduler_configs.second,
+                                                                                draft_device,
+                                                                                draft_properties,
+                                                                                false);
+    m_perf_metrics = ov::genai::SDPerModelsPerfMetrics();
+    m_perf_metrics.raw_metrics.m_inference_durations = {{MicroSeconds(0.0f)}};
+    m_draft_pipeline->raw_perf_metrics.m_inference_durations = {{ MicroSeconds(0.0f) }};
+
+    // specific params update for eagle pipeline
+    // check draft_model, retrieve d2t table if exists
+    auto d2t_tensor = extract_d2t_mapping_table(draft_model);
+    update_eagle_pipeline_params(d2t_tensor);
+}
+
+ov::Tensor ContinuousBatchingPipeline::Eagle3DecodingImpl::create_draft_input_ids(const ov::Tensor& original_input_ids) {
+    auto shape = original_input_ids.get_shape();
+    if (shape.size() == 0u || shape.back() <= 1u) {
+        return ov::Tensor(original_input_ids);
+    }
+
+    size_t original_length = shape.back();
+    size_t new_length = original_length - 1;
+
+    ov::Tensor draft_input_ids(ov::element::i64, {1, new_length});
+
+    const int64_t* src_data = original_input_ids.data<int64_t>();
+    int64_t* dst_data = draft_input_ids.data<int64_t>();
+
```

Comment:
please leave a comment - why the first token is skipped

Reply:
to match the algo in https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L683
add a comment in code as well